packages/block-editor/src/store/reducer.js

```diff
@@ -2131,6 +2131,170 @@ const combinedReducers = combineReducers( {
+				}
+			} );
+		}
+	}
+	return patternBlockEditingModes;
+}
+
+const withPatternBlockEditingModes = ( reducer ) => {
+	return ( state, action ) => {
+		const nextState = reducer( state, action );
+
+		if ( nextState === state ) {
+			return state;
+		}
+
+		nextState.patternClientIds = state?.patternClientIds ?? new Set();
+		nextState.patternBlockEditingModes =
+			state?.patternBlockEditingModes ?? new Map();
+
+		switch ( action.type ) {
+			case 'RESET_BLOCKS': {
+				nextState.patternClientIds = new Set();
+				recurseBlocks( action.blocks, ( block ) => {
+					if ( block.name === 'core/block' ) {
+						nextState.patternClientIds.add( block.clientId );
+					}
+				} );
+				nextState.patternBlockEditingModes =
+					getPatternBlockEditingModes( nextState );
+				break;
+			}
+			case 'INSERT_BLOCKS': {
+				recurseBlocks( action.blocks, ( block ) => {
+					if ( block.name === 'core/block' ) {
+						nextState.patternClientIds.add( block.clientId );
+					}
+				} );
+				nextState.patternBlockEditingModes =
+					getPatternBlockEditingModes( nextState );
+				break;
+			}
+			case 'REMOVE_BLOCKS': {
+				for ( const removedClientId of action.clientIds ) {
+					const removedTree =
+						state.blocks.tree.get(
+							`controlled||${ removedClientId }`
+						) ?? state.blocks.tree.get( removedClientId );
+					if ( removedTree ) {
+						recurseBlocks( [ removedTree ], ( block ) => {
+							nextState.patternClientIds.delete( block.clientId );
+						} );
+					}
+				}
+				nextState.patternBlockEditingModes =
+					getPatternBlockEditingModes( nextState );
+				break;
+			}
+			case 'REPLACE_INNER_BLOCKS': {
+				const rootTree =
+					state.blocks.tree.get(
+						`controlled||${ action.rootClientId }`
+					) ?? state.blocks.tree.get( action.rootClientId );
+
+				if ( rootTree ) {
+					recurseBlocks( rootTree?.innerBlocks, ( block ) => {
+						nextState.patternClientIds.delete( block.clientId );
+					} );
+				}
+				recurseBlocks( action.blocks, ( block ) => {
+					if ( block.name === 'core/block' ) {
+						nextState.patternClientIds.add( block.clientId );
+					}
+				} );
+				nextState.patternBlockEditingModes =
+					getPatternBlockEditingModes( nextState );
+				break;
+			}
+			case 'REPLACE_BLOCKS': {
+				for ( const removedClientId of action.clientIds ) {
+					const removedTree =
+						state.blocks.tree.get(
+							`controlled||${ removedClientId }`
+						) ?? state.blocks.tree.get( removedClientId );
+					if ( removedTree ) {
+						recurseBlocks( [ removedTree ], ( block ) => {
+							nextState.patternClientIds.delete( block.clientId );
+						} );
+					}
+				}
+				recurseBlocks( action.blocks, ( block ) => {
+					if ( block.name === 'core/block' ) {
+						nextState.patternClientIds.add( block.clientId );
+					}
+				} );
+				nextState.patternBlockEditingModes =
+					getPatternBlockEditingModes( nextState );
+				break;
+			}
+		}
+
+		return nextState;
+	};
+};
```

Comment:
Do you think this is the kind of code that deserves some unit testing?

Reply:
I can do, though I think the other PR (https://github.com/WordPress/gutenberg/pull/66919) shows that unit testing only the reducer or only the selectors isn't a great approach. That PR is a refactor, the store as a whole should be working the same, but I'll have to rewrite all the tests, which is a lot of work and could introduce some margin of error.

Reply:
I should also add some e2e tests.